test/pthread/test_pthread_memory_growth.c

```diff
@@ -4,42 +4,64 @@
 // found in the LICENSE file.
 
 #include <pthread.h>
-#include <sys/types.h>
-#include <stdio.h>
 #include <stdlib.h>
 #include <assert.h>
-#include <emscripten.h>
+#include <emscripten/em_js.h>
+#include <stdbool.h>
+
+// We want to test that JavaScript access in the main thread automatically updates the memory views if the heap has grown from the pthread meanwhile.
+//
+// Checking this correctly is somewhat tricky because a lot of Emscripten APIs access heap on their own, so the memory views might be updated
+// by them before we explicitly check the state in our own JS routines below, which leads to false positives (test passing even though in
+// isolation the heap access in our own JS is not the one updating the memory views).
+//
+// To test it in isolation, we need to use only EM_JS - which is as close to a pure JS call as we can get - and not EM_ASM, pthread_join, proxying
+// etc., not even `puts`/`printf` for logging - as all of those access the heap from JS on their own and might update the memory views too early.
+// Not using standard proxying mechanisms also means we need to drop down all the way to raw atomics.
+
+EM_JS(void, assert_initial_heap_state, (bool isWorker), {
+  console.log(`Checking initial heap state on the ${isWorker ? 'worker' : 'main'} thread`);
+  assert(HEAP8.length === 32 * 1024 * 1024, "start at 32MB");
+});
+
+EM_JS(void, assert_final_heap_state, (bool isWorker, const char* buffer, int finalHeapSize), {
+  console.log(`Checking final heap state on the ${isWorker ? 'worker' : 'main'} thread`);
+  assert(HEAP8.length > finalHeapSize, "end with >64MB");
+  assert(HEAP8[buffer] === 42, "readable from JS");
+});
+
+#define FINAL_HEAP_SIZE (64 * 1024 * 1024)
 
 static void *thread_start(void *arg)
 {
+  assert_initial_heap_state(true);
   // allocate more memory than we currently have, forcing a growth
-  printf("thread_start\n");
-  char* buffer = (char*)malloc(64 * 1024 * 1024);
+  char* buffer = malloc(FINAL_HEAP_SIZE);
   assert(buffer);
+  // Write value at the end of the buffer to check that any thread can access addresses beyond the initial heap size.
+  buffer += FINAL_HEAP_SIZE - 1;
   *buffer = 42;
-  pthread_exit((void*)buffer);
+  *(const char *_Atomic *)arg = buffer;
+  assert_final_heap_state(true, buffer, FINAL_HEAP_SIZE);
+  return NULL;
 }
 
 int main()
 {
-  printf("prep\n");
+  assert_initial_heap_state(false);
 
   pthread_t thr;
+  const char *_Atomic buffer = NULL;
```

Comment:
Maybe just make this a global atomic to avoid passing it via void* and the extra cast in thread_start?

Reply:
Yeah could do, either is fine by me.